lib/compat/wordpress-6.3/footnotes.php

```diff
@@ -0,0 +1,25 @@
+<?php
+/**
+ * Compatiblity filters for improved footnotes support.
+ *
+ * In core, this could be fixed directly in key functions.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Trims footnote anchors from content.
+ *
+ * @param string $content HTML content.
+ * @return string Filtered content.
+ */
+function gutenberg_trim_footnotes( $content ) {
+	if ( ! doing_filter( 'get_the_excerpt' ) ) {
+		return $content;
+	}
+
+	static $footnote_pattern = '_<sup data-fn="[^"]+" class="[^"]+">\s*<a href="[^"]+" id="[^"]+">\d+</a>\s*</sup>_';
```

Comment:
This is working pretty well, thanks! I'm not sure what else we could do here.

Looks like it handles most of the inline changes we can make to the footnote sup

<img width="503" alt="Screenshot 2023-07-12 at 1 33 43 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/1b202117-233f-4178-8217-7abcb2e2a397">

I only think with regexes like these, especially for new functionality, it would be great to have unit tests. That way when breaking changes come down the line we can react immediately.

Reply:
Also tested in Chinese and Cyrillic scripts just to make sure no i18n funny business was messing with things.

<img width="395" alt="Screenshot 2023-07-12 at 1 43 34 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8eb466be-a2cf-4f5b-bc76-c7c135b2748b">

Looks good!

Reply:
Could we use the HTML tag processor here? Cc @dmsnell

Reply:
Also interesting to know this. 

When doing the [backport](https://github.com/WordPress/wordpress-develop/pull/4845) I had a quick look at whether it could remove tags like it could with attributes/classes and I couldn't find anything. 

The other option would be a DOM parser to strip the `<sup />` and its contents but not sure if that would be a faster wa to do it.

Reply:
Is there any other possible tag inside of here besides `<sup>` and `<a>`?
The Tag Processor cannot easily remove this pattern, but the HTML Processor would be able to if I added `<sup>` support and the ability to remove a tag.

This could all plausibly happen for the WordPress 6.4 release, no promises.

Reply:
Also there's an "ugly" way to do it with the Tag Processor by subclassing and accessing internals, but the HTML Processor is working well enough I'm not experimenting with that approach anymore.

Reply:
> Could we use the HTML tag processor here?

I don't believe so because `WP_HTML_Tag_Processor` doesn't have a `remove_tag` method. If it did, you could do something like this:

```php
$p = new WP_HTML_Tag_Processor( $html );
while ( $p->next_tag( array( 'tag_name' => 'SUP' ) ) ) {
	if (
		$p->get_attribute( 'data-fn' ) &&
		$p->get_attribute( 'class' )
	) {
		$p->set_bookmark( 'footer_sup' );
		if (
			$p->next_tag() &&
			'A' === $p->get_tag() &&
			$p->get_attribute( 'href' ) &&
			$p->get_attribute( 'id' )
		) {
			$p->seek( 'footer_sup' );
			$p->remove_tag(); // 👈 This method doesn't exist.
		}
		$p->release_bookmark( 'footer_sup' );
	}
}
```

But @dmsnell would know better 😄

Reply:
(Oops race condition)

Reply:
I'm just flagging this with @dmsnell as another place we're processing HTML 😄